arangod/IResearch/Geo.cpp

```diff
@@ -21,91 +21,63 @@
 /// @author Andrey Abramov
 ////////////////////////////////////////////////////////////////////////////////
 
-#include "Geo.h"
-
+#include "IResearch/Geo.h"
+#include "IResearch/IResearchCommon.h"
 #include "Geo/GeoJson.h"
 #include "Geo/ShapeContainer.h"
 #include "Logger/LogMacros.h"
-#include "IResearch/IResearchCommon.h"
-#include "velocypack/Builder.h"
-#include "velocypack/Slice.h"
 
-namespace arangodb {
-namespace iresearch {
+#include <s2/s2latlng.h>
 
-bool parseShape(VPackSlice slice, geo::ShapeContainer& shape, bool onlyPoint) {
-  Result res;
-  if (slice.isObject()) {
-    if (onlyPoint) {
-      S2LatLng ll;
-      res = geo::geojson::parsePoint(slice, ll);
+#include <velocypack/Builder.h>
+#include <velocypack/Slice.h>
+
+namespace arangodb::iresearch {
 
-      if (res.ok()) {
-        shape.resetCoordinates(ll);
+template<bool Validation>
+bool parseShape(velocypack::Slice vpack, geo::ShapeContainer& region,
+                bool onlyPoint) {
+  Result r;
+  if (vpack.isObject()) {
+    if (onlyPoint) {
+      S2LatLng latLng;
+      r = geo::json::parsePoint<Validation>(vpack, latLng);
+      if (!Validation || r.ok()) {
+        region.reset(latLng.ToPoint());
       }
     } else {
-      res = geo::geojson::parseRegion(slice, shape, false);
+      r = geo::json::parseRegion<Validation>(vpack, region,
+                                             /*legacy=*/false);
     }
-  } else if (slice.isArray() && slice.length() >= 2) {
-    res = shape.parseCoordinates(slice, /*geoJson*/ true);
+  } else if (vpack.isArray()) {
+    r = geo::json::parseCoordinates<Validation>(vpack, region,
+                                                /*geoJson=*/true);
   } else {
-    LOG_TOPIC("4449c", DEBUG, arangodb::iresearch::TOPIC)
+    LOG_TOPIC("4449c", DEBUG, TOPIC)
         << "Geo JSON or array of coordinates expected, got '"
-        << slice.typeName() << "'";
-
+        << vpack.typeName() << "'";
     return false;
   }
-
-  if (res.fail()) {
-    LOG_TOPIC("4549c", DEBUG, arangodb::iresearch::TOPIC)
+  if (Validation && r.fail()) {
```

Comment:
Same here? Is it expected for "parseXXX" without Validation  to return failure?

Reply:
Not but in such case will be runtime condition, here it's compile time